app/components/socket-provider.tsx

```diff
@@ -0,0 +1,61 @@
+"use client";
+
+import { createContext, useContext, useEffect, useState } from "react";
+import { io } from "socket.io-client";
+import type { SocketClient } from "@/types";
+
+const socket: SocketClient = io({ autoConnect: false });
+
+const SocketContext = createContext({
+  socket,
+  isDead: false,
+});
+
+export function useSocket() {
+  return useContext(SocketContext);
+}
+
+export function SocketProvider({ children }: React.PropsWithChildren) {
+  const [isDead, setIsDead] = useState(false);
+
+  useEffect(() => {
+    const sessionID = localStorage.getItem("sessionID");
+
+    if (sessionID) {
+      socket.auth = { sessionID };
+    }
+
+    socket.connect();
+
+    function onDisconnect() {
+      console.log("disconnected");
+    }
+
+    socket.on("disconnect", onDisconnect);
+
+    socket.on("session", ({ sessionID }) => {
+      socket.auth = { sessionID };
+      localStorage.setItem("sessionID", sessionID);
+    });
+
+    socket.on("connect_error", (error) => {
+      if (error.message === "dead") {
+        setIsDead(true);
+      }
+    });
+
+    socket.on("death", () => {
+      setIsDead(true);
+    });
+
+    return () => {
+      socket.off("disconnect", onDisconnect);
+    };
+  }, []);
+
+  return (
+    <SocketContext.Provider value={{ socket, isDead }}>
+      {children}
+    </SocketContext.Provider>
+  );
```

Comment:
we can get rid of the extra `console.log`s for now, maybe we can look into socket server logging in development either with debug or middleware
see: [Socket.IO | Logging and debugging](https://socket.io/docs/v4/logging-and-debugging/)

also, I'm not the biggest fan of `isDead`, we can just use the same value that we store in redis
also also, I'm wondering if we should persist the loading state if we're still trying to retrieve the session state (so we don't get layout shift or flashbanged)
also also also, I'm tempted to rename the `"session"` and `"death"` events to be clearer, maybe `"session_alive"` and `"session_dead"`
I've also reordered the event listeners here for my own sanity

```ts
  const [sessionState, setSessionState] = useState<"alive" | "dead" | null>(null);

  useEffect(() => {
    socket.auth = { sessionId: localStorage.getItem("sessionId") };
    socket.connect();

    socket.on("connect_error", (error) => {
      if (error.message === "dead") {
        setSessionState("dead");
      }
    });

    socket.on("session_alive", (sessionId) => {
      setSessionState("alive");
      socket.auth = { sessionId };
      localStorage.setItem("sessionId", sessionId);
    });
    
    socket.on("session_dead", () => {
      setSessionState("dead");
    });
  }, []);

  return (
    <SocketContext.Provider value={{ socket, sessionState }}>
      {children}
    </SocketContext.Provider>
  );
```